src/loaders/jobs/taskfarmer/task_generator.py

```diff
@@ -0,0 +1,266 @@
+
+    return image_str
+
+
+def _create_shifter_wrapper(job_dir, image_str):
+    """
+    Creates the Shifter wrapper script.
+    """
+
+    # The content of the Shifter wrapper script
+    shifter_wrapper = f'''#!/bin/bash
+
+image={image_str}
+
+if [ $# -lt 1 ]; then
+    echo "Error: Missing command argument."
+    echo "Usage: shifter_wrapper.sh your-command-arguments"
+    exit 1
+fi
+
+command="$@"
+echo "Running shifter --image=$image $command"
+
+cd {job_dir}
+shifter --image=$image $command
+            '''
+
+    wrapper_file = os.path.join(job_dir, 'shifter_wrapper.sh')
+    with open(wrapper_file, "w") as f:
+        f.write(shifter_wrapper)
+
+    os.chmod(wrapper_file, 0o777)
+
+    return wrapper_file
+
+
+def _create_genome_id_file(genome_ids, genome_id_file):
+    """
+    Create a tab-separated values (TSV) file with a list of genome IDs.
+    """
+    with open(genome_id_file, 'w') as f:
+        f.write("genome_id\n")
+        for genome_id in genome_ids:
+            f.write(f"{genome_id}\n")
+
+
+def _create_task_list(source_data_dir, kbase_collection, load_ver, tool, wrapper_file, job_dir, root_dir):
+    """
+    Create task list file (tasks.txt)
+    """
+    genome_ids = [path for path in os.listdir(source_data_dir) if
+                  os.path.isdir(os.path.join(source_data_dir, path))]
+
+    chunk_size = CHUNK_SIZE[tool]
+    genome_ids_chunks = [genome_ids[i: i + chunk_size] for i in range(0, len(genome_ids), chunk_size)]
+
+    task_list = '#!/usr/bin/env bash\n'
+    for idx, genome_ids_chunk in enumerate(genome_ids_chunks):
+        task_list += wrapper_file
+
+        genome_id_file = os.path.join(job_dir, f'genome_id_{idx}.tsv')
+        _create_genome_id_file(genome_ids_chunk, genome_id_file)
+
+        task_list += f''' --volume={GTDBTK_DATA_PATH}:/gtdbtk_reference_data '''
+        task_list += f'''--env TOOLS={tool} --env LOAD_VER={load_ver} --env SOURCE_DATA_DIR={source_data_dir} '''
+        task_list += f'''--env KBASE_COLLECTION={kbase_collection} --env ROOT_DIR={root_dir} '''
+        task_list += f'''--env NODE_ID=job_{idx} --env GENOME_ID_FILE={genome_id_file} '''
+        task_list += f'''--entrypoint\n'''
+
+    task_list_file = os.path.join(job_dir, 'tasks.txt')
+    with open(task_list_file, "w") as f:
+        f.write(task_list)
+
+    return task_list_file, len(genome_ids_chunks)
+
+
+def _create_batch_script(job_dir, task_list_file, n_jobs):
+    """
+    Create the batch script (submit_taskfarmer.sl) for job submission
+    """
+    batch_script = '''#!/bin/sh\n'''
+    batch_script += f'''#SBATCH -N {n_jobs + 1} -c 64\n'''
+    batch_script += '''#SBATCH -q regular\n'''
+    batch_script += '''#SBATCH --time=4:00:00\n'''
+    batch_script += '''#SBATCH --time-min=0:30:00\n'''
+    batch_script += '''#SBATCH -C cpu\n\n'''
+    batch_script += f'''cd {job_dir}\n'''
+    batch_script += '''export THREADS=32\n\n'''
+    batch_script += f'''runcommands.sh {task_list_file}'''
```

Comment:
Another place for a single variable assignment with `'''` strings

Reply:
👍

Reply:
:+1: